truffle/com.oracle.truffle.tools/src/com/oracle/truffle/tools/TruffleProfiler.java

```diff
@@ -0,0 +1,433 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.truffle.tools;
+
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.frame.FrameDescriptor;
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.FrameSlotKind;
+import com.oracle.truffle.api.frame.FrameSlotTypeException;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.instrumentation.EventContext;
+import com.oracle.truffle.api.instrumentation.EventNode;
+import com.oracle.truffle.api.instrumentation.EventNodeFactory;
+import com.oracle.truffle.api.instrumentation.TruffleInstrument;
+import com.oracle.truffle.api.instrumentation.TruffleInstrument.Registration;
+import com.oracle.truffle.api.instrumentation.InstrumentationTag;
+import com.oracle.truffle.api.instrumentation.Instrumenter;
+import com.oracle.truffle.api.instrumentation.SourceSectionFilter;
+import com.oracle.truffle.api.instrumentation.SourceSectionFilter.Builder;
+import com.oracle.truffle.api.nodes.NodeCost;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import com.oracle.truffle.api.source.SourceSection;
+import com.oracle.truffle.tools.TruffleProfiler.Counter.TimeKind;
+
+@Registration(id = TruffleProfiler.ID, autostart = true)
+public class TruffleProfiler extends TruffleInstrument {
+
+    public static final String ID = "truffle_profiler";
+    private static final int MAX_CODE_LENGTH = 30;
+
+    private final Map<SourceSection, Counter> counters = new HashMap<>();
+
+    private String[] mimeTypes = parseMimeTypes(System.getProperty("truffle.profiling.includeMimeTypes"));
+
+    private boolean timingDisabled = Boolean.getBoolean("truffle.profiling.timingDisabled");
+
+    @SuppressWarnings("unused") private String includeSources = ""; // TODO implement
+    @SuppressWarnings("unused") private String excludeSources = ""; // TODO implement
+
+    // TODO temporary solution until TruffleRuntime#getCallerFrame() is fast
+    // I am aware that this is not thread safe.
+    private Counter activeCounter;
+
+    // to ensure printing by a shutdown hook
+    private boolean disposed;
+
+    @Override
+    protected void onCreate(final Env env, Instrumenter instrumenter) {
+        if (!isEnabled()) {
```

Comment:
Do I read this correctly, also looking at the execution stack when `onCreate(..)` is called, each instrument decides itself whether it is going to be active or not?

I as a language implementer would like to have a say in that.
And I as a language user would like to have a say in that as well.
Also, I might want to have different settings per `PolyglotEngine`.
The later point is especially relevant for me for testing.

The current situation, i.e., that the instrument implementer is the one who has a say about that does not seem to be ideal. Also, this means each instrument implementer will decide on their own way of dealing with this. Making it for me as language implementer or user hard to control any of this.